org.osgi.test.common/src/main/java/org/osgi/test/common/event/EventRecording.java

```diff
@@ -0,0 +1,193 @@
+		return timedEvents(ServiceEvent.class);
+	}
+
+	default List<BundleEvent> bundleEvents() {
+
+		return events(BundleEvent.class);
+	}
+
+	default List<TimedEvent<BundleEvent>> timedBundleEvents() {
+		return timedEvents(BundleEvent.class);
+	}
+
+	default List<FrameworkEvent> frameworkEvents() {
+		return events(FrameworkEvent.class);
+	}
+
+	default List<TimedEvent<FrameworkEvent>> timedFrameworkEvents() {
+		return timedEvents(FrameworkEvent.class);
+	}
+
+	static class EventsRecordingImpl implements EventRecording {
+		private List<TimedEvent<?>>	events;
+		private boolean				timedOut;
+
+		private EventsRecordingImpl() {}
+
+		public EventsRecordingImpl(boolean timedOut, List<TimedEvent<?>> events) {
+			this();
+			this.timedOut = timedOut;
+			this.events = events;
+		}
+
+		@Override
+		public boolean isTimedOut() {
+			return timedOut;
+		}
+
+		@Override
+		public List<TimedEvent<?>> timedEvents() {
+			return events;
+		}
+
+		@Override
+		public String toString() {
+			return "EventsSnapshotImpl [events=" + events + ", timedOut=" + timedOut + "]";
+		}
+
+	}
+
+	static class TimedEvent<T> {
+		private T event;
+
+		public T getEvent() {
+			return event;
+		}
+
+		public Instant getInstant() {
+			return instant;
+		}
+
+		private Instant instant = Instant.now();
+
+		public TimedEvent(T event) {
+			this.event = event;
+		}
+
+		@Override
+		public String toString() {
+			return "TimedEvent [event=" + event + ", instant=" + instant + "]";
+		}
+
+	}
+
+	static EventRecording record(BundleContext bc, Runnable action, Predicate<Object> predicate, int timeout)
+		throws Exception {
+		List<TimedEvent<?>> events = Collections.synchronizedList(new ArrayList<>());
+		CountDownLatch latch = new CountDownLatch(1);
+
+		BundleListener bListener = new SynchronousBundleListener() {
+
+			@Override
+			public void bundleChanged(BundleEvent event) {
+				events.add(new TimedEvent<BundleEvent>(CloneUtil.clone(event)));
+				if (predicate.test(event)) {
+					latch.countDown();
+				}
+
+			}
+		};
+
+		FrameworkListener fListener = new FrameworkListener() {
+
+			@Override
+			public void frameworkEvent(FrameworkEvent event) {
+				events.add(new TimedEvent<FrameworkEvent>(CloneUtil.clone(event)));
+				if (predicate.test(event)) {
+					latch.countDown();
+				}
+			}
+		};
+
+		ServiceListener sListener = new ServiceListener() {
+			@Override
+			public void serviceChanged(ServiceEvent event) {
+				events.add(new TimedEvent<ServiceEvent>(CloneUtil.clone(event)));
+				if (predicate.test(event)) {
+					latch.countDown();
+				}
+			}
+		};
+
+		bc.addFrameworkListener(fListener);
+		bc.addBundleListener(bListener);
+		bc.addServiceListener(sListener);
+		action.run();
+		boolean timedOut = !latch.await(timeout, TimeUnit.MILLISECONDS);
+		bc.removeServiceListener(sListener);
+		bc.removeBundleListener(bListener);
+		bc.removeFrameworkListener(fListener);
+
+		List<TimedEvent<?>> objects = Collections.unmodifiableList(events);
```

Comment:
You probably want to wrap the original list rather than the synchronized wrapping.

Reply:
thx. changed